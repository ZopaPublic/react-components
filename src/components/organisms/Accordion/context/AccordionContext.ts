import { createContext } from 'react';

import {
  GetAccordionHeaderProps,
  GetAccordionSectionProps,
  IsActiveAccordionSection,
} from '../../../../hooks/useAccordion/useAccordion';

export interface AccordionContext {
  getHeaderProps: GetAccordionHeaderProps;
  getSectionProps: GetAccordionSectionProps;
  isActiveSection: IsActiveAccordionSection;
}

export const AccordionContext = createContext<AccordionContext | undefined>(undefined);
