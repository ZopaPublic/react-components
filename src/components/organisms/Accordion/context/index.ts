export { AccordionContext } from './AccordionContext';
export { useAccordionContext } from './useAccordionContext';
