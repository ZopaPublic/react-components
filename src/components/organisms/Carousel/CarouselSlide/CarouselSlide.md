Slide component to be used inside `<Carousel />`
