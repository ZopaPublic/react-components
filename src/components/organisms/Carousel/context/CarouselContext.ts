import { createContext, Dispatch, SetStateAction, useContext } from 'react';

export interface CarouselContext {
  slidesCount: number;
  activeSlide: number;
  sliderHeight: number;
  setSliderHeight: Dispatch<SetStateAction<number>>;
}

const initialContext: CarouselContext = {
  sliderHeight: 0,
  activeSlide: 0,
  slidesCount: 0,
  setSliderHeight: () => console.error('Carousel context not initialized yet.'),
};

export const CarouselContext = createContext<CarouselContext>(initialContext);

export const useCarouselContext = () => useContext(CarouselContext);
