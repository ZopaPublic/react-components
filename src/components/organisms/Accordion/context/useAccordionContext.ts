import { useContext } from 'react';

import { AccordionContext } from './AccordionContext';

export const useAccordionContext = () => {
  const context = useContext(AccordionContext);
  if (context === undefined) {
    throw new Error('useAccordionContext must be used within an Accordion');
  }
  return context;
};
