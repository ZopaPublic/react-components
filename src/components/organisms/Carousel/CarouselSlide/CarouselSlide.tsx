import React, { createRef, useEffect } from 'react';
import styled, { css } from 'styled-components';
import { grid } from '../../../../constants';
import { useViewport } from '../../../../hooks/useViewport/useViewport';
import FlexCol, { FlexColProps } from '../../../layout/FlexCol/FlexCol';
import { useCarouselContext } from '../context/CarouselContext';

const StyledFlexCol = styled(FlexCol)<FlexColProps & { hidden: boolean; sliderHeight: number }>`
  text-align: center;
  position: relative;
  opacity: 1;
  max-width: 250px;
  min-height: ${({ sliderHeight }) => `${sliderHeight}px`};
  ${({ hidden }) => {
    if (hidden) {
      return css`
        position: absolute;
        opacity: 0;
      `;
    }
  }}
`;

export type SlideProps = {
  children?: React.ReactNode;
  index?: number;
};

const Slide = ({ index, children, ...rest }: SlideProps) => {
  const { slidesCount, activeSlide, sliderHeight, setSliderHeight } = useCarouselContext();
  const { width } = useViewport();
  const isActive = index === activeSlide;
  const isSmall = width && width < grid.breakpoints.m;
  const slideRef = createRef<HTMLDivElement>();

  useEffect(() => {
    const slideHeight = slideRef.current?.clientHeight || 0;
    setSliderHeight((height) => Math.max(height, slideHeight));
  }, [slideRef.current, activeSlide]);

  return (
    <StyledFlexCol
      hidden={!!isSmall && !isActive}
      sliderHeight={sliderHeight}
      xs={slidesCount}
      s={slidesCount}
      m={1}
      align="flex-start"
      {...rest}
    >
      <div ref={slideRef}>{children}</div>
    </StyledFlexCol>
  );
};

export default Slide;
