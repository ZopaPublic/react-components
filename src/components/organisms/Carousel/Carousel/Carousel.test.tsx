import { fireEvent, render, screen } from '@testing-library/react';
import axe from '../../../../../axe-helper';
import React from 'react';
import { act } from 'react-dom/test-utils';
import { Carousel } from '..';
import { SlideProps } from '../CarouselSlide/CarouselSlide';
import { useCarouselContext } from '../context/CarouselContext';

const TestSlide = ({ index }: SlideProps) => {
  const { activeSlide } = useCarouselContext();
  return <div data-automation={String(index)}>{activeSlide === index ? 'ACTIVE' : 'INACTIVE'}</div>;
};

describe('<Carousel />', () => {
  it('updates active slide when clicking next arrow', () => {
    const { getByTestId } = render(
      <Carousel>
        <TestSlide />
        <TestSlide />
        <TestSlide />
      </Carousel>,
    );
    const nextArrow = screen.getByTitle('next');

    expect(getByTestId('0').innerHTML).toEqual('INACTIVE');
    expect(getByTestId('1').innerHTML).toEqual('ACTIVE');
    expect(getByTestId('2').innerHTML).toEqual('INACTIVE');

    act(() => {
      fireEvent.click(nextArrow);
    });

    expect(getByTestId('0').innerHTML).toEqual('INACTIVE');
    expect(getByTestId('1').innerHTML).toEqual('INACTIVE');
    expect(getByTestId('2').innerHTML).toEqual('ACTIVE');

    act(() => {
      fireEvent.click(nextArrow);
    });

    expect(getByTestId('0').innerHTML).toEqual('ACTIVE');
    expect(getByTestId('1').innerHTML).toEqual('INACTIVE');
    expect(getByTestId('2').innerHTML).toEqual('INACTIVE');

    act(() => {
      fireEvent.click(nextArrow);
    });
  });

  it('updates active slide when clicking previous arrow', () => {
    const { getByTestId } = render(
      <Carousel>
        <TestSlide />
        <TestSlide />
        <TestSlide />
      </Carousel>,
    );
    const previousArrow = screen.getByTitle('previous');

    expect(getByTestId('0').innerHTML).toEqual('INACTIVE');
    expect(getByTestId('1').innerHTML).toEqual('ACTIVE');
    expect(getByTestId('2').innerHTML).toEqual('INACTIVE');

    act(() => {
      fireEvent.click(previousArrow);
    });

    expect(getByTestId('0').innerHTML).toEqual('ACTIVE');
    expect(getByTestId('1').innerHTML).toEqual('INACTIVE');
    expect(getByTestId('2').innerHTML).toEqual('INACTIVE');

    act(() => {
      fireEvent.click(previousArrow);
    });

    expect(getByTestId('0').innerHTML).toEqual('INACTIVE');
    expect(getByTestId('1').innerHTML).toEqual('INACTIVE');
    expect(getByTestId('2').innerHTML).toEqual('ACTIVE');
  });

  it('renders with no a11y violations', async () => {
    const { container } = render(
      <Carousel>
        <TestSlide />
        <TestSlide />
        <TestSlide />
      </Carousel>,
    );
    const results = await axe(container.innerHTML);
    expect(results).toHaveNoViolations();
    expect(container).toMatchSnapshot();
  });
});
