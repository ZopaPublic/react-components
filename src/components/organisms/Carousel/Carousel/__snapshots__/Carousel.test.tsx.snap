// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Carousel /> renders with no a11y violations 1`] = `
.c3 {
  background-color: transparent;
  font-size: inherit;
  font-family: "Open Sans",Roboto,Helvetica,Arial,sans-serif;
  font-weight: 600;
  line-height: inherit;
  color: #3B46C4;
  cursor: pointer;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  -webkit-user-select: auto;
  -moz-user-select: auto;
  -ms-user-select: auto;
  user-select: auto;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  padding: 0;
  border: none;
  font-family: "Open Sans",Roboto,Helvetica,Arial,sans-serif;
  font-weight: 600;
  color: auto;
}

.c3:hover,
.c3:active {
  color: #21247F;
}

.c3:hover svg path,
.c3:active svg path {
  fill: #21247F;
}

.c3:hover {
  color: auto;
}

.c3:hover svg path {
  fill: auto;
}

.c3:active {
  color: auto;
}

.c3:active svg path {
  fill: auto;
}

.c2 {
  position: relative;
  width: 100%;
  min-height: 1px;
  padding-right: 12px;
  padding-left: 12px;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
}

.c4 {
  position: relative;
  width: 100%;
  min-height: 1px;
  padding-right: 12px;
  padding-left: 12px;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
}

.c0 {
  width: 100%;
  max-width: 100%;
  padding-right: 0px;
  padding-left: 0px;
  margin-left: auto;
  margin-right: auto;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  margin-right: -12px;
  margin-left: -12px;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  margin-right: -12px;
  margin-left: -12px;
  -webkit-box-pack: space-around;
  -webkit-justify-content: space-around;
  -ms-flex-pack: space-around;
  justify-content: space-around;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
}

@media (min-width:0px) {
  .c2 {
    display: block;
    -webkit-flex: 0 0 8.333333333333332%;
    -ms-flex: 0 0 8.333333333333332%;
    flex: 0 0 8.333333333333332%;
    max-width: 8.333333333333332%;
  }
}

@media (min-width:576px) {
  .c2 {
    display: block;
    -webkit-flex: 0 0 8.333333333333332%;
    -ms-flex: 0 0 8.333333333333332%;
    flex: 0 0 8.333333333333332%;
    max-width: 8.333333333333332%;
  }
}

@media (min-width:768px) {
  .c2 {
    display: none;
  }
}

@media (min-width:0px) {
  .c4 {
    display: block;
    -webkit-flex: 0 0 83.33333333333334%;
    -ms-flex: 0 0 83.33333333333334%;
    flex: 0 0 83.33333333333334%;
    max-width: 83.33333333333334%;
  }
}

@media (min-width:576px) {
  .c4 {
    display: block;
    -webkit-flex: 0 0 83.33333333333334%;
    -ms-flex: 0 0 83.33333333333334%;
    flex: 0 0 83.33333333333334%;
    max-width: 83.33333333333334%;
  }
}

@media (min-width:768px) {
  .c4 {
    display: block;
    -webkit-flex: 0 0 100%;
    -ms-flex: 0 0 100%;
    flex: 0 0 100%;
    max-width: 100%;
  }
}

@media (min-width:576px) {
  .c0 {
    max-width: 540px;
  }
}

@media (min-width:768px) {
  .c0 {
    max-width: 720px;
  }
}

@media (min-width:992px) {
  .c0 {
    max-width: 960px;
  }
}

@media (min-width:1300px) {
  .c0 {
    max-width: 1224px;
  }
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
      cols="12"
      direction="row"
    >
      <div
        class="c2"
        cols="12"
        style="text-align: right;"
      >
        <a
          class="c3"
          color="#3B46C4"
          title="previous"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-chevron-left fa-w-10 "
            color="#818F9B"
            data-icon="chevron-left"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 320 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M34.52 239.03L228.87 44.69c9.37-9.37 24.57-9.37 33.94 0l22.67 22.67c9.36 9.36 9.37 24.52.04 33.9L131.49 256l154.02 154.75c9.34 9.38 9.32 24.54-.04 33.9l-22.67 22.67c-9.37 9.37-24.57 9.37-33.94 0L34.52 272.97c-9.37-9.37-9.37-24.57 0-33.94z"
              fill="currentColor"
            />
          </svg>
        </a>
      </div>
      <div
        class="c4"
        cols="12"
      >
        <div
          class="c0"
        >
          <div
            class="c5"
            cols="3"
            direction="row"
          >
            <div
              data-automation="0"
            >
              INACTIVE
            </div>
            <div
              data-automation="1"
            >
              ACTIVE
            </div>
            <div
              data-automation="2"
            >
              INACTIVE
            </div>
          </div>
        </div>
      </div>
      <div
        class="c2"
        cols="12"
      >
        <a
          class="c3"
          color="#3B46C4"
          title="next"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-chevron-right fa-w-10 "
            color="#818F9B"
            data-icon="chevron-right"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 320 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M285.476 272.971L91.132 467.314c-9.373 9.373-24.569 9.373-33.941 0l-22.667-22.667c-9.357-9.357-9.375-24.522-.04-33.901L188.505 256 34.484 101.255c-9.335-9.379-9.317-24.544.04-33.901l22.667-22.667c9.373-9.373 24.569-9.373 33.941 0L285.475 239.03c9.373 9.372 9.373 24.568.001 33.941z"
              fill="currentColor"
            />
          </svg>
        </a>
      </div>
    </div>
  </div>
</div>
`;
